src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/latest.sql

```diff
@@ -106,6 +106,19 @@ BEGIN
     -- restore citus catalog tables
     --
     INSERT INTO pg_catalog.pg_dist_partition SELECT * FROM public.pg_dist_partition;
+
+    -- if we are upgrading from PG14/PG15 to PG16+,
+    -- we need to add varnullingrels to partkey if it's not already there
+    IF substring(current_Setting('server_version'), '\d+')::int >= 16 THEN
+    EXECUTE $cmd$
```

Comment:
it is slightly concerning to update the pg_dist_partition like this... I mean, I cannot see anything wrong now, but I'm also not very happy about this.

One idea that I'm considering is to fully rely on PG. Use PG to store the column name before the upgrade, and then later, use the column name to set back the column in the new PG version.

On citus_prepare_pg_upgrade, we can use `column_to_column_name` to record the column names for tables:

```
CREATE TABLE public.pg_dist_partkeys_pre_16 AS SELECT logicalrelid, column_to_column_name(partkey) as col_name FROM pg_dist_partition WHERE partkey IS NOT NULL AND partkey NOT ILIKE '%varnullingrels%';; 
```

and then here:

UPDATE pg_dist_partition SET partkey = column_name_to_column(col_name) FROM pg_dist_partkeys_pre_16 WHERE pg_dist_parkeys.logicalrelid = pg_dist_partition.logicalrelid;

what do you think about that?

Reply:
I see your point.
Let me try. Maybe I don't need to put extra conditions either and just apply this in upgrades.

Reply:
Check out the successful tests in this commit https://github.com/citusdata/citus/pull/6952/commits/342c146100007eefd1a69c329b1830de848a8743